Report Autofill issues via Passwords screen

amddg44 opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.

⚠️ If you're an external contributor, please file an issue first before working on a PR, as we can't guarantee that we will accept your changes if they haven't been discussed ahead of time. Thanks!
-->

Task/Issue URL: https://app.asana.com/0/1201462886803403/1207504090894693/f
Tech Design URL:
CC:

**Description**:
Introduces new UI to allow users to report autofill breakages. The option to send a report will be visible if a user has sites included in the `Suggested` section of the Passwords screen and has not already recently submitted a report for this site. 

This feature is controlled via the new privacy config key `autofillBreakageReporter`, which also defines how many days until the option to submit another report for the same site is displayed again.

<!--
If at any point it isn't actively being worked on/ready for review/otherwise moving forward strongly consider closing it (or not opening it in the first place). If you decide not to close it, use Draft PR while work is still in progress or use `DO NOT MERGE` label to clarify the PRs state and comment with more information.
-->

**Steps to test this PR**:
Note: I’ve added a debug menu option to clear sent reports via Debug > Autofill Settings > Reset Autofill Broken Reports
1. Update the privacy config to use this URL: https://jsonblob.com/api/12628340582738739202 which adds this to the config 
```
"autofillBreakageReporter": {
      "state": “enabled”,
      "settings": {
        "monitorIntervalDays": 42
      },
      "exceptions": [
          {
            "domain": "netflix.com"              
          }
      ]
    },
```
2. Save a login for `fill.dev`
3. Visit `fill.dev` and the open the Passwords screen via the tab bar overflow menu
4. Confirm the pixel `autofill.logins.report.available` is fired
5. Confirm the option to send an anonymous report is included in the Suggested section
6. Close and access the Passwords screen via the Settings screen and confirm it is **not** visible and that the pixel `autofill.logins.report.available` is not fired
7. Go back to the Passwords screen via the overflow menu.
8. Tap `Send Anonymous Report` and confirm the pixel `autofill.logins.report.confirmation.prompt.displayed` is fired when the alert appears
9. Tap `Cancel` and confirm the pixel `autofill.logins.report.confirmation.prompt.dismissed` is fired
10. Tap `Send Anonymous Report` and confirm the toast appears and the option to send a report is removed from the table
11. Confirm both the pixels `autofill.logins.report.confirmation.prompt.confirmed` and `autofill.logins.report.failure` are fired
12. Confirm the `autofill.logins.report.failure` pixel parameter data matches your app config e.g. `email_protection` is / is not enabled, website URL is correct (but excludes query params) etc
13. Exit out of the Passwords screen, then come back via the overflow menu and confirm the option to send a report is not shown
14. Save a login for another site, and confirm that the send a report option does display for it
15. Create a login for `netflix.com` and confirm there is no option to send a report for it (as it is in the privacy config exceptions list)
16. Try doing a search for a saved password and confirm everything still works correctly 
17. Delete a few saved logins (both suggested and not) and confirm the report option is removed correctly from the UI when the `Suggested` section is empty


<!—
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
—>

**Definition of Done (Internal Only)**:

* [ ] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `’`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 15
* [ ] iOS 16
* [ ] iOS 17

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

—
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

amddg44 committed:
New autofill breakage report table view cell

amddg44 committed:
Support for breakage reporting implemented into Logins screen

amddg44 committed:
Implementation of sending acutal failure pixel with all params

amddg44 committed:
Set BSK

amddg44 committed:
Debug option to clear sent reports

amddg44 committed:
Debug option to clear sent reports

amddg44 committed:
Temporary config override for ship review build

amddg44 committed:
Minor styling tweaks

amddg44 committed:
Confirmation toast

amddg44 committed:
Manually set showBreakageReporter to false on report submitted

amddg44 committed:
Temporary hack for ship review

amddg44 committed:
Final pixel implemented

amddg44 committed:
Add missing ? to copy

amddg44 committed:
Merge remote-tracking branch 'origin/main' into anya/autofill-feedback-on-failure

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved
#	DuckDuckGo/TabViewController.swift

amddg44 committed:
Update BSK

amddg44 committed:
Tests added

amddg44 committed:
Undo ship review config hacks

amddg44 committed:
Remove accidental new line

amddg44 committed:
Change cell selection style behaviour

amddg44 committed:
Copy updates from Ship review feedback

graeme commented on DuckDuckGo/TabViewController.swift:
```diff
@@ -1852,6 +1853,34 @@ extension TabViewController: WKNavigationDelegate {
         dismiss(animated: true)
     }
+
+    private func registerForAutofillNotifications() {
```
As this class is already a bit of a beast, do you think it'd be possible to extract the breakage reporting into another type?

graeme reviewed (approved):
I've got one minor suggestion but otherwise this all looks good!

amddg44 committed:
Additional copy update from Ship review

amddg44 committed:
Merge remote-tracking branch 'origin/main' into anya/autofill-feedback-on-failure

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
BSK bump

amddg44 committed:
Autofill feedback prompt translations (#3148)

Task/Issue URL: https://app.asana.com/0/1201645688934642/1207853187721194/f
Tech Design URL:
CC:

Description:
Translations for autofill feedback prompt

amddg44 committed:
Merge remote-tracking branch 'origin/main' into anya/autofill-feedback-on-failure

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved

amddg44 committed:
Set BSK release
